lib/Dialect/SV/Transforms/HWMemSimImpl.cpp

```diff
@@ -210,14 +220,163 @@ void HWMemSimImplPass::generateMemory(HWModuleOp op, FirMemory mem) {
   outputOp->setOperands(outputs);
 }
 
+// Get all the hierarchichal names for a HWModuleOp.
+void static getHierarchichalNames(HWModuleOp op,
+                                  SmallVector<std::string> &names,
+                                  const std::string path,
+                                  mlir::SymbolUserMap &symbolUsers) {
+
+  if (!op)
+    return;
+  bool parentFound = false;
+  for (auto u : symbolUsers.getUsers(op)) {
+    if (auto inst = dyn_cast<InstanceOp>(u)) {
+      auto opPath = inst.instanceName().str() + "." + path;
+      getHierarchichalNames(inst->getParentOfType<HWModuleOp>(), names, opPath,
+                            symbolUsers);
+      parentFound = true;
+    }
+  }
+  if (!parentFound)
+    names.push_back(op.getName().str() + "." + path);
+}
+
 void HWMemSimImplPass::runOnOperation() {
-  auto topModule = getOperation().getBody();
+  auto topModule = getOperation();
+  mlir::SymbolTableCollection symbolTable;
+  mlir::SymbolUserMap symbolUsers(symbolTable, topModule);
 
   SmallVector<HWModuleGeneratedOp> toErase;
   bool anythingChanged = false;
+  DenseMap<Operation *, SmallVector<Operation *, 8>> genToMemMap;
+  SmallDenseMap<Operation *, bool> testBenchOps;
+  llvm::SetVector<Operation *> dutOps;
+  llvm::SetVector<Operation *> tbOps;
+  Attribute tbMetadataFile, dutMetadataFile, memConfAttr;
+  for (auto op : topModule.getBody()->getOps<HWModuleGeneratedOp>()) {
+    auto hwModule = cast<HWModuleGeneratedOp>(op);
+    genToMemMap[op] = {};
+    for (auto u : symbolUsers.getUsers(hwModule))
+      if (auto inst = dyn_cast<InstanceOp>(u)) {
+        genToMemMap[op].push_back(inst);
+        if (inst->hasAttr(testbenchMemAttrName)) {
+          dutMetadataFile = inst->getAttr(testbenchMemAttrName);
+          dutOps.insert(hwModule);
+        } else if (inst->hasAttr(dutMemoryAttrName)) {
+          tbMetadataFile = inst->getAttr(dutMemoryAttrName);
+          tbOps.insert(hwModule);
+        }
+        if (inst->hasAttr(memConfAttrName))
+          memConfAttr = inst->getAttr(memConfAttrName);
+      }
+  }
 
-  for (auto op :
-       llvm::make_early_inc_range(topModule->getOps<HWModuleGeneratedOp>())) {
+  auto genJson = [&](llvm::json::OStream &J, HWModuleGeneratedOp hwModule,
+                     std::string &seqMemConfStr, unsigned index,
+                     unsigned confIndex) {
+    auto mem = analyzeMemOp(hwModule);
+    std::string portStr;
+    if (mem.numWritePorts)
+      portStr += "mwrite";
+    if (mem.numReadPorts) {
+      if (!portStr.empty())
+        portStr += ",";
+      portStr += "read";
+    }
+    if (mem.numReadWritePorts)
+      portStr = "mrw";
```

Comment:
This could use a `SmallString<14>` to handle every case without a heap allocation.

Is it guaranteed that if there is a readwrite port, that there are no read or write ports? (This realtes to the lack of a check on a comma in the last if block.)